specs/packagemanager/PackageManagement.md

```diff
@@ -0,0 +1,1136 @@
+* `PackageManager.RemovePackageByFullNameAsyn(p)` fails if the specified package isn't found.
+  `PackageDeploymentManager` succeeds as the requested package is not present at the end of the
+  operation.
+  * This follows the core deployment principle "'Tis not the journey that matters but the
+    destination". In short, it doesn't matter the current state of the system, only the desired end
+    state is achieved.
+* `PackageManager` methods accept inconsistent (and often inconvenient) permutations of expressing a
+  target package. For example, `PackageManager` supports removing a package by PackageFullName but
+  not PackageFamilyName. `PackageDeploymentManager` provides a richer API accepting additional
+  identifiers.
+
+# 4. Examples
+
+## 4.1. AddPackageAsync()
+
+Fabrikam app installing Contoso's Muffin package from a .msix file.
+
+```c#
+void Install()
+{
+    var package = "d:\\contoso\\muffin.msix";
+    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
+    var options = new AddPackageOptions();
+    var deploymentResult = await packageDeploymentManager.AddPackageAsync(package, options);
+    try
+    {
+        if (deplymentResult.Status == PackageDeploymentStatus.CompletedSuccess)
+        {
+            Console.WriteLine("OK");
+        }
+        else
+        {
+            Console.WriteLine("Error {}", deploymentResult.ExtendedError());
+        }
+    }
+    catch (Exception ex)
+    {
+        Console.WriteLine(ex.ToString());
+    }
+}
+```
+
+## 4.2. AddPackageByUriAsync()
+
+Fabrikam app installing Contoso's Muffin package from an https: source.
+
+```c#
+void Install()
+{
+    var package = new Uri("https://contoson.com/muffin.msix");
+    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
+    var options = new AddPackageOptions();
+    var deploymentResult = await packageDeploymentManager.AddPackageByUriAsync(package, options);
+    try
+    {
+        if (deplymentResult.Status == PackageDeploymentStatus.CompletedSuccess)
+        {
+            Console.WriteLine("OK");
+        }
+        else
+        {
+            Console.WriteLine("Error {}", deploymentResult.ExtendedError());
+        }
+    }
+    catch (Exception ex)
+    {
+        Console.WriteLine(ex.ToString());
+    }
+}
+```
+
+**NOTE:** This differs from the AddPackageAsync() example by the method name
+(`AddPackageByUriAsync()` instead of `AddPackageAsync()`) and the target type (https: source as URI
+instead of filename as string).
+
+## 4.3. AddPackageSetAsync()
+
+Fabrikam app installing Contoso's Muffin and Waffle packages via a PackageSet.
+
+```c#
+var Install()
+{
+    var packageSet = new PackageSet();
+    var muffin = new PackageSetItem();
+    muffin.PackageUri = new Uri("c:\\contoso\\muffin-1.2.3.4.msix");
+    packageSet.Add(muffin);
+    var waffle = new PackageSetItem();
+    waffle.PackageUri = new Uri("https://contoso.com/waffle-2.4.6.8.msix");
+    packageSet.Add(waffle);
```

Comment:
How would this work with dependencies?  E.g. if I have package A which depends on frameworks X and Y, and package B which depends on frameworks Y and Z, does this get specified as a set {A, B} with dependencies {X, Y, Z}?  Or is the dependency info part of each PackageSetItem?

Reply:
Those properties are part of the operation's options (e.g. AddPackageOptions); they can't vary per `PackageSetItem` today.

I'd considered adding such things as optional override properties in `PackageSetItem` e.g.

```
    var packageSet = new PackageSet();
    var muffin = new PackageSetItem();
    muffin.PackageUri = new Uri("c:\\contoso\\muffin-1.2.3.4.msix");
    packageSet.Add(muffin);
    var waffle = new PackageSetItem();
    waffle.PackageUri = new Uri("https://contoso.com/waffle-2.4.6.8.msix");
    waffle.Stuff = "foo";
    packageSet.Add(waffle);

    var packageDeploymentManager = PackageDeploymentManager.GetDefault();
    var options = new AddPackageOptions();
    options.Stuff = "bar";
    var deploymentResult = await packageDeploymentManager.AddPackageByUriAsync(packageSet, options);
```

so muffin would be processed with Stuff=bar vs waffle is processed with Stuff=foo

But that brings a host of complexity and questions to the table e.g. PackageSet can be used in multiple contexts but not all support the same options (AddPackageOptions vs StagePackageOptions vs...). Should PackageSetItem have all the options as optional properties (packagesetitem.AddPackageOptions, .StagePackageOptions, ...)? PackageSetItem has all the individual properties as optional (packagesetitem.DevelopmentMode=true, ...)?

We have no urgent need for this variability of data, and we can easily extend the API to support it later when the need/usage is more clearly defined/understood.

TL;DR Out of scope today. Maybe in a future enhancement.